pilot/pkg/model/virtualservice.go

```diff
@@ -34,30 +34,31 @@ import (
 // This function is used by sidecar converter.
 func SelectVirtualServices(vsidx virtualServiceIndex, configNamespace string, hostsByNamespace map[string]hostClassification) []config.Config {
 	importedVirtualServices := make([]config.Config, 0)
+	vsset := sets.New[types.NamespacedName]()
 
-	vsset := sets.New[string]()
 	addVirtualService := func(vs config.Config, hosts hostClassification) {
-		vsname := vs.Name + "/" + vs.Namespace
+		key := types.NamespacedName{Namespace: vs.Namespace, Name: vs.Name}
```

Comment:
nit: `config.NamespacedName(vs)`

Reply:
using `config.NamespacedName(vs)` doubles allocations and worsens execution times for some reason, might be related to the method taking an interface and using method calls for building the struct?

might be worth checking if this is the case in other uses of this method and maybe improve or dump it?

Reply:
Interesting ,good find

Reply:
@howardjohn apparently has to do with the method taking an interface, if a create a different method that takes `Config` and does the same method calls, then no overhead is added, but when receiving `Namer` then we get a big performance hit.

maybe we can separate in two methods?
- `config.NamespacedName(config.Config)` for internal config type
- `config.NamespacedNameForObject(metav1.Object)` for k8s types

Reply:
or maybe just a method on the `Config` struct `cfg.NamespacedName()` and make the current method only take k8s objects

Reply:
```
package main

import (
	"istio.io/istio/pkg/config"
	kubetypes "k8s.io/apimachinery/pkg/types"
	"testing"
)

func Benchmark(b *testing.B) {
	obj := config.Config{
		Meta: config.Meta{
			Name:      "foo",
			Namespace: "bar",
		},
		Spec:   nil,
		Status: nil,
	}
	b.Run("Interface", func(b *testing.B) {

		for n := 0; n < b.N; n++ {
			NamespacedNameInterface(obj)
		}
	})
	b.Run("Generic", func(b *testing.B) {
		for n := 0; n < b.N; n++ {
			NamespacedNameGeneric(obj)
		}
	})
	b.Run("Special", func(b *testing.B) {
		for n := 0; n < b.N; n++ {
			NamespacedNameSpecial(obj)
		}
	})
}

type Namer interface {
	GetName() string
	GetNamespace() string
}

func NamespacedNameInterface(n Namer) kubetypes.NamespacedName {
	return kubetypes.NamespacedName{
		Namespace: n.GetNamespace(),
		Name:      n.GetName(),
	}
}

func NamespacedNameGeneric[T Namer](n T) kubetypes.NamespacedName {
	return kubetypes.NamespacedName{
		Namespace: n.GetNamespace(),
		Name:      n.GetName(),
	}
}

func NamespacedNameSpecial(n config.Config) kubetypes.NamespacedName {
	return kubetypes.NamespacedName{
		Namespace: n.Namespace,
		Name:      n.Name,
	}
}
```

```
goos: linux
goarch: amd64
pkg: istio.io/istio
cpu: 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz
Benchmark
Benchmark/Interface
Benchmark/Interface-8            7959892               172.0 ns/op           240 B/op          1 allocs/op
Benchmark/Generic
Benchmark/Generic-8             43958732                23.98 ns/op            0 B/op          0 allocs/op
Benchmark/Special
Benchmark/Special-8             1000000000               0.3507 ns/op          0 B/op          0 allocs/op
```

Seems like just making it generic is good enough. The difference between specialized and generic is pretty interesting though..

```
// 1ns
func NamespacedNameGenericSpecial1[T Namer](n T) kubetypes.NamespacedName {
	return kubetypes.NamespacedName{
	}
}
// 20ns
func NamespacedNameGenericSpecial2[T Namer](n T) kubetypes.NamespacedName {
	return kubetypes.NamespacedName{
	}
	return kubetypes.NamespacedName{
		Namespace: n.GetNamespace(),
		Name:      n.GetName(),
	}
}
```

Adding deadcode adds time

Reply:
so, do you prefer making it generic? generics have a slight performance impact, that's probably what you're seeing.

i just pushed a version where I created a `NamespacedName()` method for `Config` struct to use that instead of the generic method, i can revert that if u want.

Reply:
so basically specialized methods cost nothing, but seems a bit hard to pay 20ns for each access to this method? it really is called very frequently, those 20ns probably end up adding a lot